includes/export/class-wc-admin-report-exporter.php

```diff
@@ -0,0 +1,185 @@
+<?php
+/**
+ * Handles reports CSV export.
+ *
+ * @package WooCommerce/Export
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * WC_Admin_Report_Exporter Class.
+ */
+class WC_Admin_Report_Exporter {
+	/**
+	 * Action hook for generating a report export.
+	 */
+	const REPORT_EXPORT_ACTION = 'wc-admin_report_export';
+
+	/**
+	 * Action scheduler group.
+	 */
+	const QUEUE_GROUP = 'wc-admin-data';
+
+	/**
+	 * Export status option name.
+	 */
+	const EXPORT_STATUS_OPTION = 'wc_admin_report_export_status';
+
+	/**
+	 * Export file download action.
+	 */
+	const DOWNLOAD_EXPORT_ACTION = 'woocommerce_admin_download_report_csv';
+
+	/**
+	 * Queue instance.
+	 *
+	 * @var WC_Queue_Interface
+	 */
+	protected static $queue = null;
+
+	/**
+	 * Get queue instance.
+	 *
+	 * @return WC_Queue_Interface
+	 */
+	public static function queue() {
+		if ( is_null( self::$queue ) ) {
+			self::$queue = WC()->queue();
+		}
+
+		return self::$queue;
+	}
+
+	/**
+	 * Set queue instance.
+	 *
+	 * @param WC_Queue_Interface $queue Queue instance.
+	 */
+	public static function set_queue( $queue ) {
+		self::$queue = $queue;
+	}
+
+	/**
+	 * Hook in action methods.
+	 */
+	public static function init() {
+		// Initialize scheduled action handlers.
+		add_action( self::REPORT_EXPORT_ACTION, array( __CLASS__, 'report_export_action' ), 10, 4 );
+		// Report download handler.
+		add_action( 'admin_init', array( __CLASS__, 'download_export_file' ) );
+	}
+
+	/**
+	 * Queue up actions for a full report export.
+	 *
+	 * @param string $export_id Unique ID for report (timestamp expected).
```

Comment:
Is there any possibility of two users, both submitting requests for CSV exports at the same time? Edgey I suppose, but should we also append a user ID just in case?

Reply:
I initially included the user ID in the filename, but later removed it - see the commit history.

I chose to remove it because it doesn't make sense to restrict the files by user, and because we're already using microseconds in the export ID: https://github.com/woocommerce/woocommerce-admin/pull/2526/files/88366dd07c25cbbcacd3bc3a7267d55a9c63a705#diff-e885446237d58d740a1cbf34d483b5b5R163